---
title: "Mistranslation analysis"
author: "Sinisa Bratulic"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  github_document
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(knitr)
library(tidyverse)
library(readxl)
library(janitor)
library(seqinr)
```


## Load the CAI data

Load E. coli codon adaptation index table 
- data is from the *seqinr* package, cited below, originally published in:
Sharp, P.M., Li, W.-H. (1987) Nucleic Acids Research, 15:1281-1295.

```{r, warning=F}
data(caitab) 
caitab <- caitab %>% rownames_to_column(var = "Codon")
caitab <- caitab %>% select(Codon, cai = ec) 
knitr::kable(list(caitab[1:16,],
           caitab[17:32,],
           caitab[33:48,],
           caitab[49:64,]),
           booktabs = TRUE, valign = 't')
citation("seqinr")
```

## Load and process sfGFP sequence

```{r, warning=F, message=F}
gfp_n <- seqinr::read.fasta(file = "data/sfGFPn.fasta")
df_seq <- data.frame(Codon = splitseq(gfp_n$sfGFP)) %>% 
  left_join(caitab %>% select(Codon, cai)) %>%
  mutate(cai_roll = zoo::rollmean(cai, k = 3, fill = NA))
df_seq$Codon_c <- lapply(df_seq$Codon, s2c)
df_seq$AA <- sapply(df_seq$Codon_c, translate)
df_seq <-df_seq %>% rowid_to_column(var = "Position") %>% 
  group_by(Codon) %>% mutate(nGFP = n()) %>% ungroup()
```

## Show CAI

Rolling mean calculated over a window of 3 codons. Rolling mean is shown as a line, and CAI as colors.
```{r, warning=F, fig.height=2}
df_seq %>% 
  mutate(Codon = str_to_upper(Codon),
         #Position = paste(Position, Codon, AA)
         ) %>%
  ggplot(aes(x = Position, fill = cai)) +
  geom_rect(ymin=0, ymax=1, aes(xmin = Position -0.5, xmax = Position + 0.5), alpha = 0.4) +
  geom_line(aes(y = cai_roll))+
  theme_bw()+
  scale_fill_distiller(palette = "PuOr")+
  scale_x_continuous(breaks = seq(0,250,20))+
  #theme(axis.text.x = element_text(hjust = 1, vjust = 1)) +
  ylab("Rolling mean")
```

## Load and process manuscript data 1

```{r, message-F, warning=F}
filepos <- "data/Sub_AA_sfGFP_SuppFig6DE.xlsx"
sheets_pos <- excel_sheets(filepos)
df_pos <- lapply(sheets_pos,function(x) read_xlsx(filepos, sheet=x)) %>%
  bind_cols() %>% 
  clean_names() %>%
  select(-sf_gfp_amino_acid_residue_n_c_10)
colnames(df_pos)[1] <- "Position"

head(df_pos)

df_pos_l <- df_pos %>%
  select(-contains("residue")) %>%
  pivot_longer(cols = -Position,
               names_to = "Variant",
               values_to = "Freq") %>%
  mutate(Species = if_else(str_detect(Variant, "ec"), "Ec", "Vc"),
         Evolved = if_else(str_detect(Variant, "wild|start"), FALSE, TRUE),
         Segment = if_else(!Evolved, 
                           "0", 
                           str_replace(Variant, "^.._s*(\\d)_.*","\\1")),
         Variant = factor(Variant)
  ) %>% distinct() %>%
  left_join(df_seq %>% select(Position, Codon, AA, cai, cai_roll, nGFP)) 
  #mutate(Freq_i = if_else(is.na(Freq), 0, Freq))
df_pos_l %>% distinct(Variant, Species, Evolved, Segment)
```

# Exploratory analysis

## Aggregate plots 

- make histograms and beeswarm plots - only show measured frequencies

```{r, warning=F}
df_pos_l %>%
  ggplot(aes(x = Freq, fill = Segment)) +
  geom_histogram(bins = 30, position = "dodge")+
  facet_grid(Segment~Species) + theme_bw() +
  xlab("Mistranslation frequency [%]") +
  ggsave(device = "svg", filename = "Figs/fig1.svg")

df_pos_l %>%
  ggplot(aes(x = Segment, y = Freq, color = Segment)) +
  ggbeeswarm::geom_beeswarm()+ 
  facet_wrap(~Species) +
  theme_bw()+
  ylab("Mistranslation frequency [%]") +
  ggsave(device = "svg", filename = "Figs/fig2.svg")

```

## Mistranslation profile vs the CAI

- only show measured rates

Calculate Spearman rank correlation for mistranslation frequency vs CAI

```{r, warning=F}
library(ggpubr)
df_pos_l %>% 
  ggplot(aes(x = cai, y = Freq)) +
  geom_jitter(alpha = 0.4, width = 0.005, shape = 21) + 
  stat_cor(method = "spearman")+
  theme_bw() +
  ggsave(device = "svg", filename = "Figs/fig3.svg")
```

Weak to no correlation.

## Mistranslation profile vs the CAI (with imputed unobserved mistranslation freqs)

What if we consider positions where no mistranslation was observed?

E.g. impute missing values based on LOD.

LOD estimated as 1/10 of the lowest observed mistranslation frequency


```{r}
(LOD <- df_pos_l %>% filter(Freq >0) %>% pull(Freq) %>% min/10)

```

Unobserved mistranslation frequencies set to LOD/10

```{r}
df_pos_l <- df_pos_l %>% mutate(Freq_l = if_else(is.na(Freq) | Freq == 0, LOD, Freq))
head(df_pos_l)
```

Calculate Spearman rank correlation for mistranslation frequency vs CAI

```{r, warning=F}
df_pos_l %>% 
  ggplot(aes(x = cai, y = Freq_l)) +
  geom_jitter(alpha = 0.3, width = 0.005, shape = 21) + 
  stat_cor(method = "spearman")+
  theme_bw()+
  ggsave(device = "svg", filename = "Figs/fig4.svg")
```


## Show rolling CAI vs mistranslation frequencies

Rolling mean of CAI rescaled to (-4,0) range for plotting.

```{r, warning=F}
df_pos_l %>% 
  mutate(cai_roll = scales::rescale(cai_roll, to = c(-4,0))) %>%
  ggplot(aes(x = Position, y = log10(Freq_l), color = Species)) +
  geom_line(aes(y = cai_roll),color = "grey60", alpha = 0.8) + 
  geom_hline(yintercept = median(scales::rescale(df_seq$cai_roll, to = c(-4,0)), na.rm = T), 
             linetype = "dashed", alpha = 0.5) + 
    geom_point(shape = 21, alpha = 0.5) + 
  theme_bw() +
  scale_color_brewer(palette = "Set1") +
  ylab("Log10 Mistranslation Frequency") +
  ggsave(device = "svg", filename = "Figs/fig6.svg")

```

Overall, these CAI vs mistranslation correlation estimates are highly biased. sfGFP sequence is codon optimized and on average has a high CAI:

```{r, warning=F, message=F,fig.width=3, fig.height=2}
df_pos_l %>% 
  distinct(Position, cai) %>%
  ggplot(aes(x = cai)) +
  geom_histogram() +
  theme_bw()+
  ggsave(device = "svg", filename = "Figs/fig5.svg")
```

## Hierachical clustering based on observed mistranslated position/rate

```{r, fig.width=8, fig.height=8}
library(tidyHeatmap)
df_h <- df_pos_l %>% filter(!is.na(Freq)) %>% 
  select(Position, Variant, Freq) %>%
  pivot_wider(names_from = "Variant", values_from = "Freq") %>%
  pivot_longer(cols = -c(Position),
               names_to = "Variant") %>%
  mutate(value =if_else(is.na(value), 0, value)) %>%
  left_join(df_pos_l %>% select(-Freq) %>% distinct())
svg("Figs/fig7.svg")
fig7<-df_h %>% mutate(Position = paste(Position, Codon, AA)) %>%
  tidyHeatmap::heatmap(.row = Position,
                     .column = Variant,
                     .value = value,
                     #transform = log10,
                     palette_value = circlize::colorRamp2(c(-3, -1.5, 0, 1.5, 3), viridis::magma(5))
) %>%
  add_tile(Species) %>%
  add_tile(Segment) %>%
  add_bar(cai)
fig7
dev.off()
fig7
```

## PCA based on observed mistranslated position/frequency

```{r}
library(FactoMineR)
library(broom)
df_t <- df_h %>% 
  filter(Position != 227) %>%
  select(Position, Variant, value) %>%
  pivot_wider(names_from = "Position", values_from = "value") %>%
  left_join(df_h %>% distinct(Variant, Species, Evolved, Segment))

pca_fit <- df_t %>% 
  select(where(is.numeric)) %>%
  scale() %>%
  prcomp()
sp<-summary(pca_fit)
pca_fit %>%
  augment(df_t) %>% # add original dataset back in
  ggplot(aes(.fittedPC1, .fittedPC2, color = Species)) + 
  geom_point(size = 1.5) +
  cowplot::theme_half_open(12) + cowplot::background_grid() +
  ggrepel::geom_text_repel(aes(label = Variant),  max.overlaps =20)+
  scale_color_brewer(palette = "Set1") +
  xlab(sprintf("PC1 (%2.1f%%)",sp$importance[2,1:2][[1]]*100)) +
  ylab(sprintf("PC2 (%2.1f%%)",sp$importance[2,1:2][[2]]*100)) +
  ggsave(device = "svg", filename = "Figs/fig8.svg")
```

## Which positions contribute the most to the separation

```{r}
library(factoextra)
var_pca <- get_pca(pca_fit)
message("Top 5 positions contributing to PC1:")
sort(var_pca$contrib[,1], decreasing = T)[1:5]
message("Top 5 positions contributing to PC2:")
sort(var_pca$contrib[,2], decreasing = T)[1:5]
```


# Mistranslation profile analysis

## Load the data
```{r, warning=F, message=F}
filemt <- "data/Sub_AA_sfGFP_SuppFig6F.xlsx"
mt_sheets <- excel_sheets(filemt)
l_mt <- lapply(mt_sheets, function(x) read_xlsx(filemt, sheet=x))
names(l_mt) <- mt_sheets
df_mt <- bind_rows(l_mt, .id = "ID") %>%
  mutate(Codon = if_else(is.na(Codon), `Codon (down)`, Codon)) %>%
  select(-contains("misin"),-contains("down")) %>%
  mutate(Variant = str_remove_all(ID, "Supp Fig.. "))

df_mt_l <- df_mt %>% select(-ID) %>%
  pivot_longer(cols = -c(Variant, Codon), 
               names_to = "AA_mis",
               values_to = "Freq")

```



## PCA analysis of mistranslation profile

Sum of mistranslation rates for each codon into an aggregate 'mistranslation profile'.

Decompose the profile by PCA.

```{r}
df_t_mt <- df_mt_l %>% 
  filter(!is.na(Freq)) %>% 
  group_by(Codon, Variant) %>% summarise(sumF = sum(Freq), .groups = "drop") %>% 
  left_join(df_seq %>% distinct(Codon, AA) %>% mutate(Codon = str_to_upper(str_replace_all(Codon, "t","u")))) %>%
  mutate(Codon = paste(Codon, AA)) %>%
  select(-AA) %>%
  pivot_wider(names_from = "Codon", values_from = "sumF", values_fill = 0) %>%
  mutate(Species = if_else(str_detect(Variant, "Ec"), "Ec", "Vc"))

pca_fitmt <- df_t_mt %>% 
  select(where(is.numeric)) %>%
  scale() %>%
  prcomp()
sp2<-summary(pca_fitmt)
pca_fitmt  %>%
  augment(df_t_mt) %>% # add original dataset back in
  ggplot(aes(.fittedPC1, .fittedPC2, color = Species)) + 
  geom_point(size = 1.5) +
  cowplot::theme_half_open(12) + cowplot::background_grid() +
  ggrepel::geom_text_repel(aes(label = Variant), max.overlaps =20, force = 15) +
  scale_color_brewer(palette = "Set1")+
  xlab(sprintf("PC1 (%2.1f%%)",sp2$importance[2,1:2][[1]]*100)) +
  ylab(sprintf("PC2 (%2.1f%%)",sp2$importance[2,1:2][[2]]*100)) + 
  ggsave(device = "svg", filename = "Figs/fig9.svg")
```

## Which codons contribute the most to separation of mistranslation profiles?

```{r}

var_pca_mt <- get_pca(pca_fitmt)
message("Top 5 Codons contributing to PC1:")
sort(var_pca_mt$contrib[,1], decreasing = T)[1:5]
message("Top 5 Codons contributing to PC2:")
sort(var_pca_mt$contrib[,2], decreasing = T)[1:5]

```

## Clustering based on mistranslation profile

```{r}
svg("Figs/fig10.svg")
fig10 <- df_t_mt %>%
  pivot_longer(cols = -c(Variant, Species), names_to = "Codon", values_to = "Freq") %>%
  tidyHeatmap::heatmap(.row = Codon,
                     .column = Variant,
                     .value = Freq,
                     #transform = log10,
                     palette_value = circlize::colorRamp2(c(-3, -1.5, 0, 1.5, 3), viridis::magma(5))
) %>%
  add_tile(Species)
fig10
dev.off()
fig10
```

